pageserver/src/tenant/timeline.rs

```diff
@@ -407,7 +415,14 @@ impl std::fmt::Display for PageReconstructError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
         match self {
             Self::Other(err) => err.fmt(f),
-            Self::NeedsDownload(_tli, _layer) => write!(f, "needs download"),
+            Self::NeedsDownload(tenant_timeline_id, layer_file_name) => {
+                write!(
+                    f,
+                    "layer {}/{} needs download",
```

Comment:
Does this get formatted in contexts where there is no tenant and timeline?

Perhaps the layer file name should be enough. I was thinking of how to ensure this, perhaps there could be some debug_assert! way of making sure the context has these...

Reply:
I see that most of the usages `?` it away to `anyhow::Error`, but in which context that gets printed I'm not sure.

Is there a way to do such `debug_assert!`? I'm interested to see any example or context checking API as a lead.
Don't plan to implement it right now, but looks interesting in general.

Reply:
Trying to understand how does `tracing_error` do it ... It is not simple, and might require installing a custom layer, which however wouldn't be so bad -- only include it in debug builds to allow us to write `debug_assert_parent_spans_define(&["tenant_id", "timeline_id"]);`.

Me, earlier:

> Does this get formatted in contexts where there is no tenant and timeline?

Formatting doesn't matter but the reporting does. Perhaps we'll also need a bit of more checking up how the spans are created and propagated all over.